openshift-201/rh201-pod-auto-scale.md

```diff
@@ -0,0 +1,187 @@
+kind: HorizontalPodAutoscaler
+metadata:
+  name: hello-world-nginx-mem-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: hello-world-nginx
+  minReplicas: 1
+  maxReplicas: 5
+  metrics:
+  - type: Resource
+    resource:
+      name: memory
+      target:
+        type: AverageValue
+        averageValue: 30Mi
+```
+
+Under metrics you can set type to AverageValue and specify averageValue memory.
+
+You can also specify Utilization in the metrics section of the v2beta2 HPA.
+
+```
+metrics: 
+  - type: Deployment
+    resource:
+      name: memory 
+      target:
+        type: Utilization 
+        averageUtilization: 50 
+```
+
+Specify averageUtilization and a target average memory utilization over all the pods, represented as a percent of requested memory. The target pods must have memory requests configured.
+
+### Memory HPA
+
+For memory-based autoscaling, memory usage must increase and decrease proportionally to the replica count. On average an increase in replica count must lead to an overall decrease in memory (working set) usage per-pod. A decrease in replica count must lead to an overall increase in per-pod memory usage.
+
+Apply the `hello-world-nginx-mem-hpa` from above to your project. Trigger a re-deploy of the load-test deployment.
+
+Describe the HPA.
+
+```
+oc describe hpa hello-world-nginx-mem-hpa
+Name:                       hello-world-nginx-mem-hpa
+Namespace:                  ad204f-dev
+Reference:                  Deployment/hello-world-nginx
+Metrics:                    ( current / target )
+  resource memory on pods:  44965888 / 30Mi
+Min replicas:               1
+Max replicas:               5
+Deployment pods:            2 current / 2 desired
+Conditions:
+  Type            Status  Reason              Message
+  ----            ------  ------              -------
+  AbleToScale     True    ReadyForNewScale    recommended size matches current size
+  ScalingActive   True    ValidMetricFound    the HPA was able to successfully calculate a replica count from memory resource
+  ScalingLimited  False   DesiredWithinRange  the desired count is within the acceptable range
+Events:
+  Type    Reason             Age                 From                       Message
+  ----    ------             ----                ----                       -------
+  Normal  SuccessfulRescale  71m                 horizontal-pod-autoscaler  New size: 3; reason: All metrics below target
+  Normal  SuccessfulRescale  65m (x12 over 70m)  horizontal-pod-autoscaler  New size: 2; reason: All metrics below target
+```
+
+**Note:**
+All pods must have resource requests configured
+
+The HPA makes a scaling decision based on the observed CPU or memory utilization values of pods in an OpenShift Container Platform cluster. Utilization values are calculated as a percentage of the resource requests of each pod. Missing resource request values can affect the optimal performance of the HPA
+
```

Comment:
Probably it's better to add a note about Apps managed by Argo CD? If user's app is under Argo CD, the HPA won't work properly because it changes the original deployment specification. --> Argo CD will try to revert to the original state.

Reply:
I think it best to keep it simple and just talk about HPA and not involve Argo CD and auto heal options.